### PR TITLE
[Transaction] Transaction buffer in memory implementation.

### PR DESCRIPTION
**Pull request by congbobo184**
## Motivation
now we transaction abort or commit transaction depend on client send all messageId, but TC time out can abort this TXN, because it won't carry the messageId, so we should defend the <txn, List<MessagId>> in transaction buffer to handle the tc send abort. 
if we implement it, client will not need to send commit with messageId.
## implement
1. When we send the first txn message, the topic will load the transaction buffer and then the buffer will replay this message
2. we will open a no durable cursor.
3. the no durable cursor start position from managedLedger properties, when the txn message send every 10000(can config) it will modify the position in managedLedger.
3. if the managedLedger properties get this is null, we will start in Position.earliest.
4. we defend a struct 
```
ConcurrentOpenHashMap<TxnID, ConcurrentOpenHashSet<PositionImpl>> txnBufferCache
```
This is for positions which txn send.
when we append the txn message to the topic log complete, we will add the position into it.
commit and abort will remove it
5. we defend a struct
```
private final ConcurrentSkipListSet<PositionImpl> positionsSort = new ConcurrentSkipListSet<>();
```
This is for replay index when every 10000 txn message send finish, we can pollFirst, it is the minimum position on going transaction.
when we append the txn message to the topic log complete, we will add the position into it.
commit and abort will remove it
6. we add the state for transaction buffer, when the buffer haven't replay finish, client cant send txn message to broker.

### Verifying this change
Add the tests for it

Does this pull request potentially affect one of the following parts:
If yes was chosen, please highlight the changes

Dependencies (does it add or upgrade a dependency): (no)
The public API: (no)
The schema: (no)
The default values of configurations: (no)
The wire protocol: (no)
The rest endpoints: (no)
The admin cli options: (no)
Anything that affects deployment: (no)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 1]
```diff
@@ -53,14 +150,36 @@ public TopicTransactionBuffer(PersistentTopic topic) {
...
+                try {
+                    PositionImpl syncPosition = positionsSort.pollFirst();
+                    if (syncPosition != null) {
+                        topic.getManagedLedger().setProperty(txnOnGoingPositionName, syncPosition.toString());
+                    }
+                } catch (ManagedLedgerException | InterruptedException exception) {
+                    log.error("Topic : [{}] Position : [{}], transaction buffer " +
+                            "sync replay position fail!", topic.getName(), position);
+                }
+                countToSyncPosition.addAndGet(defaultCountToSyncPosition);
```
Comment (zymap) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:181]: I only see the `countToSyncPosition` is increased but doesn't see it decrease. That seems weird.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 2]
```diff
@@ -69,47 +188,71 @@ public TopicTransactionBuffer(PersistentTopic topic) {
...
+                if (e != null) {
+                    log.error("Failed to commit for txn {}", txnID, e);
+                    completableFuture.completeExceptionally(e);
+                    return;
+                }
+                txnBufferCache.remove(txnID);
+                completableFuture.complete(null);
+                positionsSort.removeAll(positions.values());
+            });
+        } else {
```
Comment (zymap) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:217]: We can commit a transaction even if there are no messages in it?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferInMemoryTest.java [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+    @Test
+    public void testTransactionBufferCommitAndAbort() throws Exception {
+        admin.topics().createNonPartitionedTopicAsync(TOPIC);
+        admin.topics().createSubscription(TOPIC, "test", MessageId.earliest);
+        PersistentTopic topic = (PersistentTopic) pulsar.getBrokerService()
+                .getTopic(TopicName.get(TOPIC).toString(), true).get().get();
+        TopicTransactionBuffer topicTransactionBuffer =
+                (TopicTransactionBuffer) topic.getTransactionBuffer(true).get();
+        for (int i = 0; i < 3; i++) {
```
Comment (zymap) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferInMemoryTest.java:74]: We have some tests using `Awaitility` to wait for some conditions. We can use that to do this.

**Review by zymap — COMMENTED** [review]
I just curious about some changes so I left some questions on it.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBufferState.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    protected boolean changeToReadyState() {
+        return (STATE_UPDATER.compareAndSet(this, State.Initializing, State.Ready));
+    }
+
+    protected boolean changeToInitializingState() {
+        return STATE_UPDATER.compareAndSet(this, State.None, State.Initializing);
+    }
+
+    protected boolean changeToCloseState() {
+        return (STATE_UPDATER.compareAndSet(this, State.Ready, State.Close)
```
[the remark this answers is not in the document]
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBufferState.java:59]: already close can not close twice.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 5]
```diff
@@ -53,14 +150,36 @@ public TopicTransactionBuffer(PersistentTopic topic) {
 
     @Override
     public CompletableFuture<Position> appendBufferToTxn(TxnID txnId, long sequenceId, ByteBuf buffer) {
+        if (!checkIfReady()) {
+            return FutureUtil.failedFuture(
```
[the remark this answers is not in the document]
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:154]: i think It makes the code clearer.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java (hunk 2 above)
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:217]: it is a bug in my code, i will fix it

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferInMemoryTest.java (hunk 3 above)
Comment (congbobo184) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferInMemoryTest.java:74]: yes, i will change  this to `Awaitility`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java (hunk 1 above)
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:181]: it is a bug in my code, i will fix it

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java (hunk 2 above)
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:217]: yes, when we receive the commit command again, in memory we don't have any positions need to write marker.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBufferState.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zymap) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBufferState.java:59]: Just realized. Should we call the `chagneToCloseState` when closing the `TopicTransactionBuffer`?

[the remark this answers is not in the document]
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBufferState.java:59]: now this method is not been used, when we use it we can think about the method how we handle this.

**Pull request closed without merging**